rfcs/0000-rlp-cleanup.md

```diff
@@ -0,0 +1,331 @@
+
+### Goals of cleaning up the RLP
+
+1. Decouple the API from the underlying implementation - i.e. provide a more generic abstraction
+2. Simplify the API - i.e. DRY, straight to the point (for defining rate limits)
+3. Consistency with Kuadrant's [AuthPolicy](https://pkg.go.dev/github.com/kuadrant/kuadrant-operator@v0.2.1/api/v1beta1#AuthPolicy) API - i.e. same language, similar UX
+
+### Current WIP to consider while discussing the cleaning up of the RLP
+
+1. Inheritance of the network matching rules from the upstream network object (https://github.com/Kuadrant/architecture/pull/4)
+2. Improvements to RateLimitPolicy - aka "minimal working example" without `configurations`, `limits.conditions` and `limits.variables` (https://github.com/Kuadrant/kuadrant-operator/issues/133)
+3. Implement `skip_if_absent` for the RequestHeaders action (https://github.com/Kuadrant/wasm-shim/issues/29)
+4. Rate Limiting across clusters ([doc](https://docs.google.com/document/d/1pqCODRAkNUTLB6lJfRWcv3d2Hlj9WqsGySmjrP707Vk/edit#heading=h.nzpgr3pef6uy))
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Given the following HTTPRoute:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: toystore
+  labels:
+    app: toystore
+spec:
+  parentRefs:
+    - name: istio-ingressgateway
+      namespace: istio-system
+  hostnames: ["*.toystore.com"]
+  rules:
+    - matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
+      backendRefs:
+        - name: toystore
+          port: 80
+```
+
+These are examples of RateLimitPolicies targeting the route:
+
+**Example 1.** Minimal example - unconditional, unqualified rate limiting
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - rates:
+        - limit: 5
+          unit: minute
+```
+
+**Example 2.** All you can get - conditions, counter qualifiers, multiple rates, and weights
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: writers
+      selectors:
+        # this is a condition
+        - selector: request.http.method
```

Comment:
So is the difference here that one has qualifiers and one doesn't. It is is an interesting idea, it did take me a minute to understand that by reading it

Reply:
@maleck13, I imagine you refer to distinction between what is a "condition" and what is a "counter qualifier". While the former includes an operator and value to be matched, the latter is just the selector.

This was one of the suggestions that came up during one of our tech discussion sections. I made sure to include other possible (including more explicit) variations here in the doc: https://github.com/Kuadrant/architecture/blob/0cc37bde1c90efad05290251635f7bd16f43d929/rfcs/0000-rlp-cleanup.md#possible-variations-for-the-selectors-conditions-and-counter-qualifiers

Reply:
I like it for the conditions, just not 100% sold on the counters. When I read selector here it makes sense that they are defining when a limit should be applied but using it for defining the counter feels a bit odd, not that I couldn't be convinced.

Reply:
@guicassolato I might have missed but are these selectors an `and` operation when there are multiple?

Reply:
They are, @maleck13